### ISSUE TITLE
Change of behavior when calling parent methods statically

### ISSUE THREAD

**Issue by theilig** [mlively/Phake#280]
Noticed this while trying to upgrade from 2.1.1 to 3.1.6

```
class behaviorChange {
    public static function tryB() {
        return static::b();
    }
    public function b() {
        return "here";
    }
}

class behaviorChangeTest {
    public function testTryB() {
        $mock = Phake::mock('behaviorChange');
        Phake::whenStatic($mock)->tryB->thenCallParent();
        Phake::whenStatic($mock)->b->thenReturn('there');
       print_r($mock::tryB());
    }
} 
```

in 2.1.1 the created Phake class called directly into the parent, which kept the late static binding associated to the mock, resulting in the static::b call correctly being mocked as well (returning 'there')

however in 3.1.6 the Phake class calles ParentDelegateCallback->__invoke()
which calls invokeArgs, which calls reflectionMethod with a null context.  This means the late static binding will no longer be the mock, but instead be the SUT.  static::b then does not hit the mock (returning 'here')

I suspect passing the mock as the context would fix it rather than null, but without looking it doesn't seem like an easy thing to do

**Comment by webkod3r** [comment]
I'm experiencing the same issue. I have a class mocking a static method, but when called internally it doesn't call the mocked one but instead it makes the actual call. Any idea or workaround?

**Comment by theilig** [comment]
here's a workaround patch, it's not ideal because it removes anything you might want from the Phake stubber callback, but it's what we've been doing for the last 3 years with success

`--- a/src/Phake/Stubber/Answers/ParentDelegate.php
+++ b/src/Phake/Stubber/Answers/ParentDelegate.php
@@ -81,14 +81,7 @@ class Phake_Stubber_Answers_ParentDelegate implements Phake_Stubber_IAnswer
 
             if (!$reflMethod->isAbstract())
             {
-                if (defined('HHVM_VERSION'))
-                {
-                    return array('parent', $method);
-                }
-                else
-                {
-                    return new Phake_Stubber_Answers_ParentDelegateCallback($context, $reflMethod);
-                }
+                return array('parent', $method);
             }
         }
         catch (ReflectionException $e)
`

**Comment by webkod3r** [comment]
Thanks @theilig, this behavior was changed in v2.2.1 and instantly broke the test my team had in their repo. Changing this will imply making changes to the vendor dependencies which I"m not a big fan of. But I guess we don't have other options. Thanks.

**Comment by theilig** [comment]
agreed, for now we have our own forked version which isn't great.  However
we have hundreds of tests that rely on this behavior.

**Comment by adoy** [comment]
I'll have a look at if/how we can configure that

**Comment by theilig** [comment]
Update to this issue, php 8.2 no longer allows 'parent' to be passed to call_user_func

Basically ParentDelegate calls only work with instances, not for static calls.  If you want late static binding you need the Phake class itself invoking parent:: (calling SomeClass__PHAKE::myFunction will lead to an infinite loop, and calling SomeClass::myFunction will not match a whenStatic on the Phake.  

In 8.2 we kept the behavior of returning ['parent'::$method] but also needed to alter the class generation template as well, to special case the return of ['parent', x] from getAnswerCallback.  

here is the additional change to MockClass
--- a/src/Phake/ClassGenerator/MockClass.php
+++ b/src/Phake/ClassGenerator/MockClass.php
@@ -443,6 +443,10 @@ class {$newClassName} {$extends}
            {
            \$__PHAKE_result = \$__PHAKE_callback(\$__PHAKE_args);
            }
+           else if (is_array(\$__PHAKE_callback) && \$__PHAKE_callback[0] === 'parent')
+           {
+           \$__PHAKE_result = parent::{$method->getName()}(...\$__PHAKE_args);
+           }
            else
            {
            \$__PHAKE_result = call_user_func_array(\$__PHAKE_callback, \$__PHAKE_args);
